google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkService.java

```diff
@@ -16,35 +16,234 @@
+ * always called on EDT thread. This class is not thread-safe, must be used on UI thread.
  */
-// TODO(ivanporty) implementation coming in the next PR
 public class ManagedCloudSdkService implements CloudSdkService {
+  private Logger logger = Logger.getInstance(ManagedCloudSdkService.class);
+
+  private ManagedCloudSdk managedCloudSdk;
+
+  private SdkStatus sdkStatus = SdkStatus.NOT_AVAILABLE;
+
+  private final Collection<SdkStatusUpdateListener> statusUpdateListeners = Lists.newArrayList();
+
+  private volatile ListenableFuture<Path> runningInstallationJob;
+
+  @Override
+  public void activate() {
+    // TODO track event that custom SDK is activated and used.
+
+    install();
+  }
 
   @Nullable
   @Override
   public Path getSdkHomePath() {
-    return null;
+    switch (sdkStatus) {
+      case READY:
+        return managedCloudSdk.getSdkHome();
+      default:
+        return null;
+    }
   }
 
   @Override
-  public void setSdkHomePath(String path) {}
+  public void setSdkHomePath(String path) {
+    /* unsupported, to be removed. */
+  }
 
   @Override
   public SdkStatus getStatus() {
-    return SdkStatus.NOT_AVAILABLE;
+    return sdkStatus;
   }
 
   @Override
-  public boolean installAutomatically() {
+  public boolean install() {
+    // check if installation is already running first, to prevent multiple jobs running.
+    if (runningInstallationJob == null || runningInstallationJob.isDone()) {
+      runningInstallationJob =
+          ThreadUtil.getInstance().executeInBackground(this::installManagedSdk);
+      Futures.addCallback(
+          runningInstallationJob,
+          new FutureCallback<Path>() {
+            @Override
+            public void onSuccess(Path path) {
+              if (path != null) {
+                logger.info("Managed Google Cloud SDK successfully installed at: " + path);
+              }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+              if (t instanceof InterruptedException) {
+                logger.info("Managed Google Cloud SDK installation cancelled.");
+              }
+            }
+          });
+    }
+
     return true;
   }
 
   @Override
-  public void addStatusUpdateListener(SdkStatusUpdateListener listener) {}
+  public void addStatusUpdateListener(SdkStatusUpdateListener listener) {
+    statusUpdateListeners.add(listener);
+  }
+
+  @Override
+  public void removeStatusUpdateListener(SdkStatusUpdateListener listener) {
+    statusUpdateListeners.remove(listener);
+  }
+
+  public void cancelInstall() {
+    if (runningInstallationJob != null) {
+      runningInstallationJob.cancel(true);
+    }
+  }
+
+  @VisibleForTesting
+  ManagedCloudSdk createManagedSdk() throws UnsupportedOsException {
+    return ManagedCloudSdk.newManagedSdk();
+  }
+
+  @VisibleForTesting
+  void setLogger(Logger logger) {
+    this.logger = logger;
+  }
+
+  @VisibleForTesting
```

Comment:
Could this (having to expose this for testing) be solved by expanding what we have in ThreadUtils? Specifically having a swappable executor that runs on the EDT?

Reply:
Yes I thought that too, it's pretty useful considering our way of testing UI related things. I searched for existing code in IJ testing framework that might do this but didn't find it. We can add something similar to ThreadUtil. How about a separate PR for that?

Reply:
yeah separate pr is fine. maybe add a todo on it then

Reply:
done.